subsys/settings/src/settings_file.c

```diff
@@ -519,16 +517,51 @@ void settings_mount_file_backend(struct settings_file *cf)
 	settings_line_io_init(read_handler, write_handler, get_len_cb, 1);
 }
 
+static int mkdir_if_not_exists(const char *path)
+{
+	struct fs_dirent entry;
+	int err;
+
+	err = fs_stat(path, &entry);
+
+	/* If directory doesn't exist, create it */
+	if (err == -ENOENT) {
+		err = fs_mkdir(path);
+	}
+
+	return err;
+}
+
+static int mkdir_for_file(const char *file_path)
+{
+	char dir_path[SETTINGS_FILE_NAME_MAX];
+	int err;
+
+	for (size_t i = 0; file_path[i] != '\0'; i++) {
+		if (i > 0 &&
+		    file_path[i] == '/') {
+			dir_path[i] = '\0';
+
+			err = mkdir_if_not_exists(dir_path);
+			if (err) {
+				return err;
+			}
+		}
+
+		dir_path[i] = file_path[i];
+	}
+
+	return 0;
+}
```

Comment:
Maybe this should be update to VFS as `fs_mkdir_path` function?
This seems useful beyond settings. Of course addition to VFS would require adding tests for this, but it is worth doing.

Reply:
Hmmm, maybe not in `fs.c` as `fs_mkdir_path()`, but rather in new `fsutil.c` as `fsutil_mkdir_path()`? It is a functionality built on top of FS layer, something high-level.

But in general I agree that this would be useful for applications. The amount of use of `mkdir -p` in shell scripts on Linux is a good advocate for such functionality. And since we don't have scripts in Zephyr, but rather C code for everything, such function would be really helpful.

Main implementation detail is how to allocate buffer for `char dir_path[]`?  In this `settings_file.c` it is trivial, as there is already maximum path restriction, which is small enough to put the buffer on stack. But how to do it in a generic way? There are some options:
- stack allocated (should be enough for 32/64 bytes long, maybe even 256 in some cases if user know what he is doing); can be configurable by Kconfig,
- static buffer (can be any size, as it does not occupy stack and is not "calling thread dependent"), which would require mutex protected access; can be configurable by Kconfig,
- heap allocated (always or only when exceeding stack allocated buffer); one such heap-dependent functionality doesn't hurt, but if we build a lot of such functionalities, then this can bite users in complex multi-threaded applications when there will be high pressure on heap.

Out of the options above, I'd go with first (stack allocated) option, with configurable buffer size in Kconfig. If we exceed that, then simply an error is returned.

I see two APIs that would be useful, equivalents of following bash commands:
```
mkdir -p /some/dir
mkdir -p `dirname /some/dir/path`
```
where both would result in creating `/some/dir`. Note that current `mkdir_for_file()` is equivalent of the latter.

Overall I would prefer to keep going with "settings" overhaul for now and not get sidetracked, at least for now. We can always replace those static functions with new generic FS utilities, right?

Reply:
> Hmmm, maybe not in `fs.c` as `fs_mkdir_path()`, but rather in new `fsutil.c` as `fsutil_mkdir_path()`? It is a functionality built on top of FS layer, something high-level.
> 

Yes, that would be right.

> Overall I would prefer to keep going with "settings" overhaul for now and not get sidetracked, at least for now. We can always replace those static functions with new generic FS utilities, right?

OK, so lets log an enhancement from the https://github.com/zephyrproject-rtos/zephyr/pull/52866/files#r1043363194 and have the PR as is.

Reply:
Added an enhancement: #52921